### PR TITLE
[new module] Azure Traffic Manager profile module

### PR DESCRIPTION
**Pull request by caohai**
##### SUMMARY

- add azure_rm_trafficmanagerprofile module
- add azure_rm_trafficmanagerprofile_facts module
- add integration test for these two modules
- add traffic manager client and models in azure_rm_common
- add necessary pkg info

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

azure_rm_trafficmanager
azure_rm_trafficmanager_facts
##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (haicao-azure-traffic-manager cfc5d2b222) last updated 2018/08/08 03:08:50 (GMT +000)
  config file = None
  configured module search path = [u'/home/hai/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/hai/ansible/lib/ansible
  executable location = /home/hai/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by caohai** [comment]
@mattclay Hi Matt, this PR cannot pass sanity test due to error in a file that I didn't touch: **[azure_rm_storageaccount.py]**. Any suggestion?

**Comment by mattclay** [comment]
@caohai The unrelated failures should be fixed now. I've restarted CI for this PR.

**Comment by mattclay** [comment]
Still one more unrelated issue to resolve. I'm working on that.

**Comment by mattclay** [comment]
The remaining sanity failure is due to changes in this PR. Something is preventing the azure_rm_storageaccount module from using the Azure SDK to enumerate the storage types, which in turn causes the validation of the module to fail.

**Comment by mattclay** [comment]
@caohai You can run the sanity tests locally so you don't have to wait for Shippable. For example, once you've done `source hacking/env-setup` you can run:

`ansible-test sanity -v lib/ansible/module_utils/azure_rm_common.py ...` to test files you've modified. You may want to add one of the following options though, depending on your environment, to provide the necessary requirements:

- `--requirements` if you're already in a virtual environment
- `--tox` if you have tox available
- `--docker` if you have docker available

If your working branch is based off of your local devel branch you can avoid having to list filenames by using the `--changed` option. This won't work if your devel branch has commits not present in your working branch though.

**Comment by mattclay** [comment]
@caohai You can also run the integration tests locally if you have access to an Azure account. For example `ansible-test integration azure_rm_trafficmanager -v --docker`. You'll need to follow the instructions in https://github.com/ansible/ansible/blob/devel/test/integration/cloud-config-azure.yml.template to configure your credentials first.

**Comment by caohai** [comment]
@mattclay Thank you for your advice, I will try them. 
I have tried to move all my model import into traffic manager module. The only thing left in azure_rm_common is TrafficMangerManagementClient, I'm not sure why import TrafficMangerManagementClient cause azure_rm_storageaccount to fail sanity test.

**Comment by mattclay** [comment]
@caohai I figured out why the sanity failure is occurring. It is due to the `azure_rm_storageaccount` module depending on having `HAS_AZURE = True` to pass validation. The docker container used for validation does not have `azure.mgmt.trafficmanager`, which means `HAS_AZURE = False`.

I need to discuss options for fixing this with a few people tomorrow and then I'll get back to you. I should know what direction we're taking tomorrow, but may not have a solution implemented by then.

**Comment by caohai** [comment]
@mattclay Thank you for the information. That's also what I have found.
I just did sanity test locally and no error occurred as expected. Hope this can get resolved soon or can we just ignore this error for now?

**Comment by yungezz** [comment]
ready_for_review

**Comment by yungezz** [comment]
@nitzmahone  @jborean93  HI Matt and Jordan, could you pls help to review this PR? this traffic manager module is in 2.7 release candidate, it's for web app scenario. Thanks in advance.

**Comment by Fred-sun** [comment]
@nitzmahone @jborean93 could you help review this PR? Thanks!

**Comment by Fred-sun** [comment]
@nitzmahone @jborean93 Thanks for your contribution, Could you help review this PR? Thanks!

**Comment by Fred-sun** [comment]
@caohai Thanks for the contribution, Could you help resolve the conflicting?

**Comment by Fred-sun** [comment]
@yungezz @caohai Could you help confirm the PR ready for review or not? Thanks!

**Comment by caohai** [comment]
ready_for_review

**Comment by Fred-sun** [comment]
@nitzmahone @jborean93  Could you help review this PR? Thanks!

**Comment by Fred-sun** [comment]
@nitzmahone @jborean93 Could you help review this PR? Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_storageaccount.py [hunk 1]
```diff
@@ -158,7 +158,10 @@ class AzureRMStorageAccount(AzureRMModuleBase):
     def __init__(self):
 
         self.module_arg_spec = dict(
-            account_type=dict(type='str', choices=[], aliases=['type']),
+            account_type=dict(
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_storageaccount.py]: You don't need to remove this but usually we separate different module fixes into different PRs.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+    state:
+        description:
+            - Assert the state of the Traffic Manager profile. Use C(present) to create or update a Traffic Manager profile and C(absent) to delete it.
+        default: present
+        choices:
+            - absent
+            - present
+    location:
+        description:
+            - Valid azure location. Defaults to 'global'.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Does this mean the location defaults to `global` if not set or defaults to `global` if only creating the profile? If it is the former it should be set as a default for the module option. If it is the latter then it should be documented as so. Is there a reason why it doesn't default to the resource group location like most of our other module do?

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+    profile_status:
+        description:
+            - The status of the Traffic Manager profile.
+        default: Enabled
+        choices:
+            - Enabled
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Use `enabled` and `disabled` for this. It's a bit of a mix up when to use snake_case over PascalCase but for a simple enumeration like this we should favour the Ansible preferred form of snake_case.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+            - Disabled
+    traffic_routing_method:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Use `routing_method`, we already know we are dealing with a Traffic Manager object so we can shorten the option.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+        description:
+            - The traffic routing method of the Traffic Manager profile.
+        default: Performance
+        choices:
+            - Performance
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Use snake_case for this, e.g. `performance`, `priority`, `weighted`. If you need to send this form to the API you can always use https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/common/dict_transformations.py#L14 and just change the first char to upper case.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+            - Weighted
+            - Geographic
+    dns_config:
+        description:
+            - The DNS settings of the Traffic Manager profile.
+        suboptions:
+            relative_name:
+                description:
+                    - The relative DNS name provided by this Traffic Manager profile.
+                    - If no provided, name of the Traffic Manager will be used
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: s/no/not/

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+                    - HTTP
+                    - HTTPS
+                    - TCP
+            port:
+                description:
+                    - The TCP port used to probe for endpoint health.
+            path:
+                description:
+                    - The path relative to the endpoint domain name used to probe for endpoint health.
+            interval_in_seconds:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Change to `interval`, add in the description that it is in seconds.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+                description:
+                    - The monitor interval for endpoints in this profile.
+            timeout_in_seconds:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Change to `timeout`, add in the description that it is in seconds.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+                description:
+                    - The monitor timeout for endpoints in this profile.
+            tolerated_number_of_failures:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Change to `tolerated_failures`, `tolerated_number_of_failures` is a bit too long.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 10]
```diff
@@ -0,0 +1,580 @@
...
+          Environment: Test
+
+    - name: Delete a Traffic Manager Profile
+      azure_rm_trafficmanager:
+        state: absent
+        name: tmtest
+        resource_group: tmt
+'''
+RETURN = '''
+state:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Just return the `id` of the traffic manager profile (and maybe a list of endpoint ids). We should only be returning things the user may not know. We use the _facts module to return all the remaining information in an easier format. So this should be

```
RETURN = '''
id:
  description: The ID of the traffic manager profile
  returned: always
  type: str
  example: "/subscriptions/XXXXXX...XXXXXXXXX/resourceGroups/tmt/providers/Microsoft.Network/trafficManagerProfiles/tmtest"
endpoints:
  description: List of endpoint IDs attached to the profile
  returned: always
  type: list
  sample: [ "id1...", "id2..." ]
'''
```

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py [hunk 11]
```diff
@@ -0,0 +1,580 @@
...
+                description:
+                    - The monitor timeout for endpoints in this profile.
+            tolerated_number_of_failures:
+                description:
+                    - The number of consecutive failed health check before declaring an endpoint in this profile Degraded after the next failed health check.
+        default:
+            protocol: HTTP
+            port: 80
+            path: /
+    endpoints:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Endpoints should be managed in another module. Have `endpoints` here take in a list of endpoint names or ids instead of the whole endpoint definition.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager_facts.py [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+        name: Testing
+        resource_group: TestRG
+
+    - name: Get facts for all Traffic Manager profiles
+      azure_rm_trafficmanager_facts:
+
+    - name: Get facts by tags
+      azure_rm_trafficmanager_facts:
+        tags:
+          - Environment:Test
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanager_facts.py]: We should also support tags being a dict where the key is the tag name and the value is the value, e.g.

```
tags:
  Environment: test
```

**Review by jborean93 — COMMENTED** [review]
Have left some comments, probably the biggest bit is separating the endpoint logic out into a different module.

File: lib/ansible/modules/cloud/azure/azure_rm_storageaccount.py (hunk 1 above)
Comment (caohai) [reply lib/ansible/modules/cloud/azure/azure_rm_storageaccount.py]: As @mattclay [mentioned](https://github.com/ansible/ansible/pull/43812#issuecomment-411986915), the docker container used for validation does not have azure.mgmt.trafficmanager. This will cause sanity test failure on [azure_rm_storageaccount](https://github.com/ansible/ansible/pull/43812#issuecomment-412044096). We made this change to azure_rm_storageaccount.py to pass CI temporarily.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager_facts.py (hunk 12 above)
Comment (caohai) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanager_facts.py]: I checked several existing Azure facts modules and it seems that support tags being a list has been a convention. One example: https://github.com/ansible/ansible/blob/4c8808ec9df095a0bf5d9252eb65a07063843bcc/lib/ansible/modules/cloud/azure/azure_rm_resourcegroup_facts.py#L53-L57

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py (hunk 2 above)
Comment (caohai) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: In default public Azure cloud, Traffic Manager profile can only be deployed globally.(ref in [doc](https://docs.microsoft.com/en-us/azure/traffic-manager/quickstart-create-traffic-manager-profile#create-a-traffic-manager-profile))
Trying to create a Traffic Manager in a location like WestUS will get an error message:
```bash
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Error creating the Traffic Manager: The subscription is not registered for the resource type 'trafficmanagerprofiles' in the location 'westus'. Please re-register for this provider in order to have access to this location."}
```
Can I state this in documentation and set a default value in module option at the same time?
Just in case other Azure cloud environment like Azure Stack, Azure China and Azure US Gov would support different Traffic Manager profile location.(Not sure about this point)

File: lib/ansible/modules/cloud/azure/azure_rm_storageaccount.py (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_storageaccount.py]: Yep that's ok, you can always raise a 2nd PR to fix the choices sanity test and then rebase this based on the changes after that. But that's not strictly necessary.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py (hunk 2 above)
Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: Fair enough, if that's the case then set the default to `global` as part of the argument spec and document it as such.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager_facts.py (hunk 12 above)
Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanager_facts.py]: Ok food for thought, I know the `tags` option for non facts modules can take in the dict format (or maybe the reverse). This should be something we try and unify in the future.

**Comment by caohai** [comment]
@jborean93 Hi Jordan, I updated the PR according to your comments and endpoints operation has been removed. Could you please review this PR again? Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+    state:
+        description:
+            - Assert the state of the Traffic Manager profile. Use C(present) to create or update a Traffic Manager profile and C(absent) to delete it.
+        default: present
+        choices:
+            - absent
+            - present
+    location:
+        description:
+            - Valid azure location. Defaults to 'global' because in default public Azure cloud, Traffic Manager profile can only be deployed globally.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py:40]: Add the key `default: global` instead of just adding it to the description. You should be setting this as part of your argument spec as well (commented below).

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+                type='str',
+                required=True
+            ),
+            state=dict(
+                type='str',
+                default='present',
+                choices=['present', 'absent']
+            ),
+            location=dict(
+                type='str'
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: Set `default='global'` here

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+    - name: Delete a Traffic Manager Profile
+      azure_rm_trafficmanager:
+        state: absent
+        name: tmtest
+        resource_group: tmt
+'''
+RETURN = '''
+id:
+    description: The ID of the traffic manager profile
+    returned: always
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: The code shows that this is only returned when `state: present`. I would change the returned key to be `traffic manager profile exists`. Update the code to always return this and the endpoints list when `state: present` and when `state: absent` and you actually removed an endpoint.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+    type: str
+    example: "/subscriptions/XXXXXX...XXXXXXXXX/resourceGroups/tmt/providers/Microsoft.Network/trafficManagerProfiles/tmtest"
+endpoints:
+  description: List of endpoint IDs attached to the profile
+  returned: always
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: Ditto to this regarding returning this in more cases and updating the docs.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 17]
```diff
@@ -0,0 +1,451 @@
...
+                if update_tags:
+                    to_be_updated = True
+
+                to_be_updated = to_be_updated or self.check_update(response)
+
+            if to_be_updated:
+                self.log("Need to Create / Update the Traffic Manager profile")
+
+                if not self.check_mode:
+                    self.results = shorten_traffic_manager_dict(self.ceate_update_traffic_manager_profile())
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: `self.results` should still contain the info on the profile even in check mode if the profile exists. This should be set if you get a get a response and then updated here in case it was updated.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 18]
```diff
@@ -0,0 +1,451 @@
...
+                    self.log("Creation / Update done.")
+
+                self.results['changed'] = True
+                return self.results
+
+        elif self.state == 'absent' and response:
+            self.log("Need to delete the Traffic Manager profile")
+            self.results['changed'] = True
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: results should also contain the id and endpoints ids of the resource that was deleted. There's not too much point in knowing this post delete but it keeps the return values more consistent.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 19]
```diff
@@ -0,0 +1,451 @@
...
+
+        :return: deserialized Traffic Manager profile state dictionary
+        '''
+        self.log("Creating / Updating the Traffic Manager profile {0}".format(self.name))
+
+        parameters = Profile(
+            tags=self.tags,
+            location=self.location,
+            profile_status=self.profile_status,
+            traffic_routing_method=self.routing_method,
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py:413]: Will this still work with the lowercase first char that was changed?

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 20]
```diff
@@ -0,0 +1,451 @@
...
+        Creates or updates a Traffic Manager profile.
+
+        :return: deserialized Traffic Manager profile state dictionary
+        '''
+        self.log("Creating / Updating the Traffic Manager profile {0}".format(self.name))
+
+        parameters = Profile(
+            tags=self.tags,
+            location=self.location,
+            profile_status=self.profile_status,
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py:412]: Will this still work with the lowercase first char that was changed?

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 21]
```diff
@@ -0,0 +1,451 @@
...
+        except CloudError as exc:
+            self.log('Error attempting to create the Traffic Manager.')
+            self.fail("Error creating the Traffic Manager: {0}".format(exc.message))
+
+    def check_update(self, response):
+        if response['location'] != self.location:
+            self.log("Location Diff - Origin {0} / Update {1}".format(response['location'], self.location))
+            return True
+
+        if response['profile_status'].lower() != self.profile_status.lower():
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: You shouldn't need the `.lower()` on the `self.profile_status`, the module arg spec will ensure it is lower case based on the choices you set.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 22]
```diff
@@ -0,0 +1,451 @@
...
+            self.log("Profile Status Diff - Origin {0} / Update {1}".format(response['profile_status'], self.profile_status))
+            return True
+
+        if response['routing_method'].lower() != self.routing_method.lower():
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: You shouldn't need the `.lower()` on the `self.routing_method`, the module arg spec will ensure it is lower case based on the choices you set.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 19 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py:413]: yes

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 20 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py:412]: yes

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 19 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py:413]: covered by test

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 20 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py:412]: covered by test

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 21 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: updated

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 22 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: updated

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 16 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: updated

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 15 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: updated

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 13 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py:40]: updated

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py (hunk 4 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py (hunk 9 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py (hunk 10 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py (hunk 11 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanager.py]: fixed

**Comment by yungezz** [comment]
hi @jborean93  comments are resolved. endpoint is separated in its own module, could you pls review it again? thanks.

**Comment by yungezz** [comment]
ready_for_review

**Comment by yungezz** [comment]
HI @jborean93  could you pls help to review the PR  again? all comments are resolved. thanks.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py [hunk 23]
```diff
@@ -0,0 +1,457 @@
...
+        if not self.location:
+            self.location = 'global'
+
+        response = self.get_traffic_manager_profile()
+
+        if self.state == 'present':
+            if not response:
+                to_be_updated = True
+            else:
+                self.results = response
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: This would return all the fields and not our curated ones, this needs to be shortened.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 24]
```diff
@@ -0,0 +1,350 @@
...
+    profile_name:
+        description: Name of Traffic Manager profile where this endpoints attaches to.
+        type: str
+        required: true
+    type:
+        description:
+            - The type of the endpoint.
+        required: true
+        choices:
+            - azureEndpoints
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: Make these snake case, `azure_endpoints`, `external_endpoints`, `nested_endpoints`. There's a module_util you can use to convert it back to this form.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 25]
```diff
@@ -0,0 +1,350 @@
...
+            - externalEndpoints
+            - nestedEndpoints
+    target_resource_id:
+        description:
+            - The Azure Resource URI of the of the endpoint.
+            - Not applicable to endpoints of type 'ExternalEndpoints'.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: This needs to be changed, make it `Not applicable to endpoints of I(type) C(external_endpoints).`

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 26]
```diff
@@ -0,0 +1,350 @@
...
+        type: str
+    target:
+        description:
+            - The fully-qualified DNS name of the endpoint.
+        type: str
+    enable:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: We are defining the state so the option should be `enabled` not `enable`

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 27]
```diff
@@ -0,0 +1,350 @@
...
+        description:
+            - The status of the endpoint.
+        type: bool
+        default: true
+    weight:
+        description:
+            - The weight of this endpoint when using the 'Weighted' traffic routing method.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: refer to the actual options in the traffic manager profile module and not just a copy/paste of the MS docs. In this case it should be something like

`The weight of the endpoint when the traffic profile has a routing_method of C(weighted)`

This has the benefit of showing both the trafficmanagerprofile method name (routing_method) and the value of it in Ansible.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 28]
```diff
@@ -0,0 +1,350 @@
...
+            - Possible values are from 1 to 1000.
+        type: int
+    priority:
+        description:
+            - The priority of this endpoint when using the 'Priority' traffic routing method.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: See the weight option comment

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 29]
```diff
@@ -0,0 +1,350 @@
...
+            - Possible values are from 1 to 1000, lower values represent higher priority.
+            - This is an optional parameter. If specified, it must be specified on all endpoints.
+            - No two endpoints can share the same priority value.
+        type: int
+    location:
+        description:
+            - Specifies the location of the external or nested endpoints when using the 'Performance' traffic routing method.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py:73]: See the weight option comment

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 30]
```diff
@@ -0,0 +1,350 @@
...
+        type: str
+    min_child_endpoints:
+        description:
+            - The minimum number of endpoints that must be available in the child profile in order for the parent profile to be considered available.
+            - Only applicable to endpoint of type 'NestedEndpoints'.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: `Only application to an endpoint of I(type) C(nested_endpoints).

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 31]
```diff
@@ -0,0 +1,350 @@
...
+        type: int
+    geo_mapping:
+        description:
+            - The list of countries/regions mapped to this endpoint when using the 'Geographic' traffic routing method.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: See the weight option comment

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 32]
```diff
@@ -0,0 +1,350 @@
...
+                type='str',
+                required=True
+            ),
+            profile_name=dict(
+                type='str',
+                required=True
+            ),
+            type=dict(
+                type='str',
+                choices=['azureEndpoints', 'externalEndpoints', 'nestedEndpoints'],
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: snake_case these

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py [hunk 33]
```diff
@@ -0,0 +1,350 @@
...
+                # delete when no exists
+                self.fail("Traffic Manager endpoint {0} not exists.".format(self.name))
+
+        if self.action == Actions.CreateOrUpdate:
+            self.results['changed'] = True
+            if self.check_mode:
+                return self.results
+
+            response = self.create_update_traffic_manager_endpoint()
+            self.results['id'] = response['id']
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py:261]: You should be documenting this.

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint_facts.py [hunk 34]
```diff
@@ -0,0 +1,283 @@
...
+        required: True
+    profile_name:
+        description:
+            - Name of Traffic Manager Profile
+        required: True
+    type:
+        description:
+            - Type of endpoint.
+        choices:
+            - azureEndpoints
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint_facts.py]: See my point in the non facts module around snake_case.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 24 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 25 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 26 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 27 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 29 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py:73]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 30 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 31 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 32 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py (hunk 33 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint.py:261]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint_facts.py (hunk 34 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerendpoint_facts.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py (hunk 23 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_trafficmanagerprofile.py]: fixed

**Comment by yungezz** [comment]
ready_for_review

**Comment by yungezz** [comment]
hi @jborean93  the pr updated to resolve comments, could you pls help to review again? Thanks a lot!

**Comment by Fred-sun** [comment]
@caohai @yungezz The PR exist conflicting files, Could you help resolve? Thanks!

**Review by jborean93 — APPROVED** [review]
